### PR TITLE
ISOM-1111: add publishing scripts

### PR DESCRIPTION
**Pull request by harishv7**
### TL;DR
Add a new script (`publisher.sh`) and supporting TypeScript code to clone the `isomer` repo, fetch site data, and build schemas.

### What changed?
- Added `publisher.sh` to handle the publishing workflow, including cloning the repository, building the components, fetching from the database, and generating the site.
- Script includes a step to run preBuild and build scripts from remote.
- Created supporting TypeScript files to fetch and write site data (navbar, footer, config) and process resources from the database.
- Added `.env.template` for environment variables.
- Added `package.json` and `package-lock.json` for dependencies.
- Configured `tsconfig.json` for TypeScript settings.

### How to test?
1. Set up the required environment variables in `.env` based on `.env.template`.
2. Run `publisher.sh` and verify that it correctly clones the repository, fetches data, and generates the site.
3. Check the output directories for schema and data files to confirm they have been created correctly.
4. Validate the logging output to ensure all steps succeeded without errors.

### Why make this change?
This script and supporting code streamline the process of building and publishing the site by automating the steps involved, thereby reducing the potential for human error and making the process more efficient.

Closes ISOM-1111 and ISOM-1245

### CODE REVIEW

**Comment by harishv7** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/421?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#421** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/421?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#382** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/382?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#381** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/381?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#369** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/369?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#365** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/365?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @harishv7 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: tooling/build/scripts/publisher.sh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  duration=$((end_time - start_time))
+  echo "Time taken: $duration seconds"
+}
+
+# Cloning the repository
+echo "Cloning repository..."
+start_time=$(date +%s)
+
+# TODO: add single branch checkout with --branch main
+git clone --depth 1 --branch 08-03-add_publishing_scripts https://github.com/opengovsg/isomer.git
```
Comment (harishv7) [inline tooling/build/scripts/publisher.sh:16]: TODO: update this before merge!

File: tooling/build/scripts/publishing/index.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+      return;
+    }
+
+    const filePath = path.join(directoryPath, fileName);
+
+    // Create directories if they don't exist
+    fs.mkdirSync(directoryPath, { recursive: true });
+
+    // Write JSON content to file
+    fs.writeFileSync(filePath, JSON.stringify(content, null, 2), "utf-8");
```
Comment (dcshzj) [inline tooling/build/scripts/publishing/index.ts:140]: nit: I think we might not need to beautify the JSON here, since it is directly consumed by Next.js.

Comment (dcshzj) [inline tooling/build/scripts/publishing/index.ts:140]: nit: I think we can combine this with `writeJsonToFile`?

File: tooling/build/scripts/publishing/queries.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        public."Resource" r
+        LEFT JOIN public."Version" v ON v."id" = r."publishedVersionId"
+        LEFT JOIN public."Blob" b ON v."blobId" = b.id
+
+        -- This join determines if the recursion continues if there are more rows
+        INNER JOIN "resourcePath" path ON r."parentId" = path.id
+    WHERE
+        r."siteId" = $1 AND (r.type = 'Folder' OR (r.type = 'Page' AND b."content" IS NOT NULL))
+    )
+    SELECT * FROM "resourcePath";
```
Comment (dcshzj) [inline tooling/build/scripts/publishing/queries.ts:42]: issue: I think this is retrieving all the pages of the site but instead it should be only retrieving pages with a `publishedVersionId`.

File: tooling/build/scripts/publisher.sh [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+# Prebuild
+echo "Prebuilding..."
+rm -rf node_modules && rm -rf .next
+curl https://raw.githubusercontent.com/opengovsg/isomer/main/tooling/build/scripts/preBuild.sh | bash
+
+# Build
+echo "Building..."
+rm -rf scripts/
+curl https://raw.githubusercontent.com/opengovsg/isomer/main/tooling/build/scripts/build.sh | bash
```
Comment (dcshzj) [inline tooling/build/scripts/publisher.sh:77]: nit: personal preference but I think we can combine the preBuild and build scripts into this one so it is easier to update/test in the future.

File: tooling/build/scripts/publishing/index.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    // Process each resource
+    for (const resource of resources) {
+      logDebug(
+        `Processing resource with id ${resource.id}, fullPermalink: ${resource.fullPermalink}`
+      );
+      if (resource.type === "Page" && resource.content) {
+        await writeContentToFile(
+          resource.fullPermalink,
+          resource.content,
```
Comment (dcshzj) [inline tooling/build/scripts/publishing/index.ts:68]: issue: I think there needs to be some adjustments to include the props that have been moved out of the `content` column (e.g. page title).

File: tooling/build/scripts/publishing/queries.ts (hunk 3 above)
Comment (harishv7) [reply tooling/build/scripts/publishing/queries.ts:42]: @dcshzj the left join is on the publishedVersionId right

```
LEFT JOIN public."Version" v ON v."id" = r."publishedVersionId"
```

Comment (dcshzj) [reply tooling/build/scripts/publishing/queries.ts:42]: Hmm yes but left join means that all rows on the left table (Resources) [would be kept](https://stackoverflow.com/a/51619897), even if the `on` condition with the NULL values is evaluated to false.

Comment (harishv7) [reply tooling/build/scripts/publishing/queries.ts:42]: Ah i see what you mean, will fix!

Comment (harishv7) [reply tooling/build/scripts/publishing/queries.ts]: I have added a where clause at the end now

File: tooling/build/scripts/publishing/index.ts (hunk 5 above)
Comment (harishv7) [reply tooling/build/scripts/publishing/index.ts:82]: Added in line 77

File: tooling/build/scripts/publishing/index.ts (hunk 2 above)
Comment (harishv7) [reply tooling/build/scripts/publishing/index.ts]: done

Comment (harishv7) [reply tooling/build/scripts/publishing/index.ts]: the logic is a bit different for writing to schema vs data folders, preferring to refactor this later if needed

File: tooling/build/scripts/publisher.sh (hunk 4 above)
Comment (harishv7) [reply tooling/build/scripts/publisher.sh]: makes sense, made the changes

**Review by dcshzj — APPROVED** [review]

**Pull request merged**